feat: Implement redesigned gas fee modal

OGPoyraz opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR aims to implement redesigned gas fee modal.

It validates all cases below:

✅ - EIP1559 transaction with `dapp-suggested` gas values
✅ - EIP1559 transaction with `fee-market` gas estimates
✅ - EIP1559 transaction with `gasPrice` estimate
✅ - EIP1559 transaction with `legacy` gas estimates - (BSC Only option)
✅ - Legacy transaction with `dapp-suggested` gas values
✅ - Legacy transaction with `fee-market` gas estimates
✅ - Legacy transaction with `gasPrice` estimate
✅ - Legacy transaction with `legacy` gas estimates - (BSC Only option)
✅ - Wallet initiated with `fee-market` gas estimates (EIP-1559)
✅ - Wallet initiated with `legacy` gas estimates (EIP-1559) - (BSC Only option)
✅ - Wallet initiated with `gasPrice` estimates (EIP-1559)

There are couple recordings added for demonstrate in the recordings section. Manual QA of this component will be done only after `transfer` confirmation is enabled.

Following tasks will be handled separately

❌ - Save values for the network - Won't do as team decided
✅ - Metrics implementation
✅ - Advanced gas modal validations

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/4831

## **Manual testing steps**

Only possible to test it out via setting FEATURE_FLAG_REDESIGNED_TRANSFER to true

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

### Wallet initiated EIP1559 transaction


https://github.com/user-attachments/assets/d548d6d6-67e9-435d-b61c-b09d08ba2790


### DApp initiated EIP1559 transaction


https://github.com/user-attachments/assets/f77c6543-6aaa-4961-afce-c64b512f235e


### DApp initiated legacy transaction


https://github.com/user-attachments/assets/2cdd237c-c4ec-4575-bd57-886e9175d5be


### DApp initiated EIP1559 without dapp suggestion



https://github.com/user-attachments/assets/13f6cb64-417c-4476-b213-52d2f50daf8d




<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

OGPoyraz committed:
Implement redesigned gas fee modal

OGPoyraz committed:
Fix lint and test

OGPoyraz committed:
Fix lint and tests

OGPoyraz committed:
Fix test

OGPoyraz committed:
Fix more unit test

OGPoyraz committed:
Fix sonar issue

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/components/gas-limit-input/gas-limit-input.test.tsx:
```diff
@@ -0,0 +1,39 @@
+
+import { GasLimitInput } from './gas-limit-input';
+
+jest.mock(
```
Minor, is this something we want to mock? I understand our goal with React unit tests was to do the minimum mocking and instead use a mock Redux store with mock state? To make them more integration-like?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/components/gas-limit-input/gas-limit-input.tsx:
```diff
@@ -0,0 +1,52 @@
+
+  const handleChange = (text: string) => {
+    setValue(text);
+    onChange(add0x(decimalToHex(text) as Hex));
```
Minor, could define a `updatedGasLimitHex` for clarity?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/components/gas-modal-header/gas-modal-header.tsx:
```diff
@@ -0,0 +1,38 @@
+  TextVariant,
+} from '../../../../../../../../component-library/components/Texts/Text';
+
+export const GasModalHeader = ({
```
Minor, is there anything gas specific about this or could this be `BackButtonHeader` for example?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/components/max-base-fee-input/max-base-fee-input.tsx:
```diff
@@ -0,0 +1,91 @@
+              ),
+            })}
+          </Text>
+          <Text variant={TextVariant.BodySM} style={styles.infoLabel}>
```
Minor, is it worth a local component to remove this duplication? Such as `Info`?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/components/max-base-fee-input/max-base-fee-input.tsx:
```diff
@@ -0,0 +1,91 @@
+  ).toString();
+  const [value, setValue] = useState(initialMaxBaseFee);
+
+  const { gasFeeEstimates } = useGasFeeEstimates(
```
Out of scope for this PR, but I'd love to call this `useNetworkGasFeeData` in future, or create an alias. To keep it distinct from the actual estimates now coming from the `TransactionMeta`.

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/constants.ts:
```diff
@@ -0,0 +1,5 @@
+export const MODALS = {
```
Would this be safer as an `enum` such as `GasModalType`?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useAdvancedGasFeeOption.ts:
```diff
@@ -0,0 +1,116 @@
+
+  const onAdvancedGasFeeClick = useCallback(() => {
+    if (transactionEnvelopeType === TransactionEnvelopeType.legacy) {
+      setActiveModal(MODALS.ADVANCED_GAS_PRICE);
```
Minor, could we instead assign a variable such as `newModalType` using a ternary? And call `setActiveModal` only once?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useAdvancedGasFeeOption.ts:
```diff
@@ -0,0 +1,116 @@
+    }
+  }, [transactionEnvelopeType, setActiveModal]);
+
+  const isCustomUserFeeLevelSelected = userFeeLevel === 'custom';
```
Can we use `UserFeeLevel.CUSTOM` instead?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useAdvancedGasFeeOption.ts:
```diff
@@ -0,0 +1,116 @@
+  const isCustomUserFeeLevelSelected = userFeeLevel === 'custom';
+
+  const isDappSuggestedGasFeeSelected = useMemo(() => !!(
+      userFeeLevel === 'dappSuggested' ||
```
Also here, can we use the `UserFeeLevel` enum?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useAdvancedGasFeeOption.ts:
```diff
@@ -0,0 +1,116 @@
+
+  const isCustomUserFeeLevelSelected = userFeeLevel === 'custom';
+
+  const isDappSuggestedGasFeeSelected = useMemo(() => !!(
```
Why do we need any of these `useMemo` if we're only setting a primitive `boolean`? It shouldn't have any impact to re-renders?

matthewwalsh0 replied:
Minor, can still see these.

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useAdvancedGasFeeOption.ts:
```diff
@@ -0,0 +1,116 @@
+  ]);
+
+  let value = '';
+  let valueInFiat = '--';
```
Minor, can we use a constant for this? Is it rendered?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useAdvancedGasFeeOption.ts:
```diff
@@ -0,0 +1,116 @@
+    valueInFiat = currentCurrencyFee || '';
+  }
+
+  return [
```
As this is a hook returning a non-primitive, this is what should be in a `useMemo` to avoid getting a fresh object each time?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useDappSuggestedGasFeeOption.ts:
```diff
@@ -0,0 +1,83 @@
+      });
+
+    options.push({
+      emoji: '🌐',
```
Minor, is it worth defining the emojis as constants together somewhere? Perhaps an enum such as `GasOptionIcon`?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useGasFeeEstimateLevelOptions.ts:
```diff
@@ -0,0 +1,172 @@
+            maxPriorityFeePerGas = transactionGasFeeEstimates?.gasPrice;
+            break;
+        }
+        gasPropertiesToUpdate = {
```
Why are we updating the gas fees directly if a level is selected? Isn't that the responsibility of the `TransactionController` now? Meaning we're duplicating all that knowledge here?

Is there a delay before this happens? Should we trigger this automatically from the `updateTransactionGasFees` method?

OGPoyraz replied:
Good question, you are right on the point that it gets delayed until next gas fee controller update hits in. Only then the `updateTransactionGasFees` callback in `packages/transaction-controller/src/helpers/GasFeePoller.ts` will be executed, after that `txParams` will be updated.

This logic seems similar but in fact it's new one. Because updates here takes `transactionMeta.gasFeeEstimates` as a source and it updates accordingly. The function mentioned above takes processed `GasFeeEstimates` from `GasFeeController`.

So similarly we need to create an updater for `TransactionController.updateTransactionGasFees` if `userFeeLevel` is changed since we want to pull this logic into controller.

OGPoyraz replied:
PR: https://github.com/MetaMask/core/pull/5800

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useGasPriceEstimateOption.ts:
```diff
@@ -0,0 +1,120 @@
+
+  const options: GasOption[] = [];
+
+  if (shouldIncludeGasPriceEstimateOption) {
```
Minor, can we early return in scenarios like this rather than nesting the core logic?

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/gas-fee-modal/modals/advanced-eip1559-modal/advanced-eip1559-modal.tsx:
```diff
@@ -0,0 +1,104 @@
+    maxPriorityFeePerGas: null,
+  });
+
+  const onSaveClick = useCallback(() => {
```
Minor, `handleSaveClick`?

pedronfigueiredo commented on app/components/Views/confirmations/components/modals/gas-fee-modal/modals/advanced-gas-price-modal/advanced-gas-price-modal.tsx:
```diff
@@ -0,0 +1,93 @@
+  });
+
+  const onSaveClick = useCallback(() => {
+    updateTransactionGasFees(transactionMeta.id, {
```
Would it be better to only submit properties that have actually changed, or that are truthy?

OGPoyraz replied:
Make sense, done.

OGPoyraz commented on app/components/Views/confirmations/components/rows/transactions/gas-fee-details/gas-fee-details.tsx:
```diff
@@ -9,12 +9,14 @@ import useHideFiatForTestnet from '../../../../../../hooks/useHideFiatForTestnet
 import styleSheet from './gas-fee-details.styles';
 
-const GasFeesDetails = () => {
+const GasFeesDetails = ({ isCompact = false }) => {
```
Do you mean `disableUpdate` ?

matthewwalsh0 replied:
I meant the inverse as it seemed more common, but either way is fine.

OGPoyraz commented on app/components/Views/confirmations/utils/time.ts:
```diff
@@ -0,0 +1,51 @@
+  },
+});
+
+export const determineEstimatedTime = (min: number, max: number) => {
```
Done - it's gas specific that's right - you have any suggestion to relocate it or?

OGPoyraz committed:
Address suggestions

OGPoyraz committed:
Fix lint

OGPoyraz committed:
Fix lint

OGPoyraz committed:
Fix test

OGPoyraz committed:
Remove unintended file change

OGPoyraz committed:
Update useGasFeeEstimateLevelOptions

OGPoyraz committed:
Remove unnecessary hook dependencies

OGPoyraz committed:
Tiny type fix

OGPoyraz committed:
Remove unused import

OGPoyraz committed:
Updates

OGPoyraz committed:
Rename prop

OGPoyraz committed:
Fix lint

OGPoyraz committed:
Reduce margin in values

OGPoyraz committed:
Fix lint

OGPoyraz committed:
Add final design adjustments

OGPoyraz commented on app/components/Views/confirmations/components/modals/gas-fee-modal/hooks/useAdvancedGasFeeOption.test.ts:
I re-evaluated this and will relocate files.

OGPoyraz replied:
Done, relocated all files to respect current structure

OGPoyraz committed:
feat: Adds gas modal metrics (#15383)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR aims to add gas modal metrics.

Fixes: https://github.com/MetaMask/mobile-planning/issues/2194

N/A

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

<!-- [screenshots/recordings] -->

<!-- [screenshots/recordings] -->

- [X] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [X] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

OGPoyraz committed:
feat: Implement gas input validations in redesigned gas modals (#15478)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR aims to implement gas input validations.

To see it in action, please see the recordings below.

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/4919

N/A

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

<!-- [screenshots/recordings] -->

https://github.com/user-attachments/assets/222338d4-8bf9-4421-b121-e66dd9a7b2c3

https://github.com/user-attachments/assets/6d7fc75e-34f0-4c80-afc6-153014b24172

- [X] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [X] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

OGPoyraz committed:
Relocate files

OGPoyraz committed:
Revert unintentional change

OGPoyraz committed:
fix final suggestions from design

OGPoyraz committed:
Fix tests

OGPoyraz committed:
Fix test

matthewwalsh0 commented on app/components/Views/confirmations/components/modals/advanced-eip1559-modal/advanced-eip1559-modal.tsx:
```diff
@@ -0,0 +1,141 @@
+
+  const handleSaveClick = useCallback(() => {
+    updateTransactionGasFees(transactionMeta.id, {
+      userFeeLevel: 'custom',
```
Minor, can we use enum here?

OGPoyraz replied:
Done

matthewwalsh0 commented on app/components/Views/confirmations/hooks/gas/useGasPriceEstimateOption.ts:
```diff
@@ -0,0 +1,124 @@
+    valueInFiat: currentCurrencyFee || '',
+  });
+
+  return options;
```
Minor, as we're returning a non-primitive, do we need a `useMemo`?

OGPoyraz replied:
Done

OGPoyraz committed:
Final updates

OGPoyraz commented:
E2e failures on smoke confirmations are expected: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/023c8e5c-d633-4884-8049-2123ee47e896

They are failing on different reason, explained here: https://consensys.slack.com/archives/C02U025CVU4/p1748480127828199

![Screenshot 2025-05-29 at 08 22 07](https://github.com/user-attachments/assets/432a7162-3c13-488e-a8f3-7aad351f694d)
